### PR TITLE
python3Packages.bip32: init at 3.4

### PR DESCRIPTION
**Pull request by arcnmx**
[bip32](https://github.com/darosior/python-bip32) is a required dependency of ledger-bitcoin 0.2.2, which was [updated from 0.2.1](https://github.com/NixOS/nixpkgs/commit/bccd14d7c855a1499f6ae198cbaf87ef47803db0) as part of #251878. It's currently failing to build.
This fixes the currently failing build as well `pkgs.electrum` which depends on it.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by oxalica** [comment]
Is there any reason to not use `fetchPypi`? [3.4](https://pypi.org/project/bip32/3.4/) is released on pypi.

**Comment by arcnmx** [comment]
> Is there any reason to not use `fetchPypi`? [3.4](https://pypi.org/project/bip32/3.4/) is released on pypi.

setuptools failed to work with the pypi distribution (it looks like `setup.py` or `requires.txt` need patching/renaming), so I figured it was easier to use the source tarball that just works as-is. I could hack around that and switch back to pypi as src though if that's preferred.

**Comment by oxalica** [comment]
> setuptools failed to work with the pypi distribution (it looks like `setup.py` or `requires.txt` need patching/renaming?),

That's fair. But it's better to comment this in the code.

File: pkgs/development/python-modules/python-bip32/default.nix [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+, fetchFromGitHub
+, buildPythonPackage
+, pytestCheckHook
+, pythonOlder
+, base58
+, coincurve
+}:
+
+buildPythonPackage rec {
+  pname = "python-bip32";
```
Comment (natsukium) [inline pkgs/development/python-modules/python-bip32/default.nix:11]:
```suggestion
  pname = "bip32";
```
pname should match the library's name on PyPI.
https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/python.section.md#contributing-guidelines-contributing-guidelines

File: pkgs/development/python-modules/python-bip32/default.nix [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+  version = "3.4";
+  format = "setuptools";
```
Comment (natsukium) [inline pkgs/development/python-modules/python-bip32/default.nix:13]:
```suggestion
  pyproject = true;
```
Could you please use `pyproject = true` instead?
https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/python.section.md#buildpythonpackage-parameters-buildpythonpackage-parameters

File: pkgs/development/python-modules/python-bip32/default.nix [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+  disabled = pythonOlder "3.9";
+
+  src = fetchFromGitHub {
+    inherit version;
```
Comment (natsukium) [inline pkgs/development/python-modules/python-bip32/default.nix:18]:
```suggestion
```
I think we don't need this.

Comment (arcnmx) [reply pkgs/development/python-modules/python-bip32/default.nix:18]: Sorry yeah, that was left over from attempting to use `fetchPypi` ><

**Comment by Yarny0** [comment]
##### Reviewed points

- [x] package path fits guidelines
- [ ] package name fits guidelines
- [ ] package version fits guidelines
- [x] package build on x86-64
- [ ] executables tested on ARCHITECTURE
- [x] `meta.description` is set and fits guidelines
- [ ] `meta.license` fits upstream license
- [ ] `meta.platforms` is set
- [x] `meta.maintainers` is set
- [x] build time only dependencies are declared in `nativeBuildInputs`
- [ ] source is fetched using the appropriate function
- [x] the list of `phases` is not overridden
- [x] when a phase (like `installPhase`) is overridden it starts with `runHook preInstall` and ends with `runHook postInstall`.
- [x] patches that are remotely available are fetched with `fetchpatch`

##### Possible improvements

*   Maybe adding the package `python-bip32` and adding the missing dependencies to `ledger-bitcoin` could be two separate commits.
*   Is the license correct?  https://github.com/darosior/python-bip32/blob/1492d39312f1d9630363c292f6ab8beb8ceb16dd/LICENCE looks like `bsd3` (or is this the same as MIT ?).

##### Comments

I tested the pull request at hand by building `electrum` with it (via `ledger-bitcoin`): It starts, and viewing transactions and accounts seems to work flawlessly (I didn't *create* any transactions though).  Due to this, I recommend this pull request to be merged to unbreak `electrum` on `master`.

(Note: This review is based on the branch `ba10d73`)

**Comment by arcnmx** [comment]
> Is the license correct?

Good catch! It seems like the [license is inaccurately listed in its metadata](https://github.com/darosior/python-bip32/blob/1492d39312f1d9630363c292f6ab8beb8ceb16dd/setup.py#L20), but it is indeed bsd3.

**Comment by arcnmx** [comment]
Rebased past #264290 to attempt to resolve the eval error.

**Comment by arcnmx** [comment]
> There is unfortunately no tag of this release available on github yet, so the commit hash is used directly in the src.

This is [no longer necessary](https://github.com/NixOS/nixpkgs/compare/38e8d70075fb03fedba4a53d06ec492582ebe481..edcac9fa079320b2bd22e826319aa8b85418f221)

**Comment by natsukium** [comment]
@ofborg build python311Packages.bip32

**Comment by natsukium** [comment]
Result of `nixpkgs-review pr 263636` at edcac9fa run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>12 packages built successfully:</summary>
  <ul>
    <li>electrum</li>
    <li>electrum-grs</li>
    <li>electrum-grs.dist</li>
    <li>electrum.dist</li>
    <li>python310Packages.bip32</li>
    <li>python310Packages.bip32.dist</li>
    <li>python310Packages.ledger-bitcoin</li>
    <li>python310Packages.ledger-bitcoin.dist</li>
    <li>python311Packages.bip32</li>
    <li>python311Packages.bip32.dist</li>
    <li>python311Packages.ledger-bitcoin</li>
    <li>python311Packages.ledger-bitcoin.dist</li>
  </ul>
</details>

**Review by natsukium — APPROVED** [review]
LGTM, thanks.

**Pull request merged**